### PR TITLE
Make Lazy.locking() thread-safe

### PR DESCRIPTION
**Pull request by big-guy**
LockingLazy relied on non-volatile fields, which can be unreliable under
heavy thread contention. This change makes the implementation thread-safe
by taking the implementation of Guava's Suppliers.memoize()

The LazyTest tests were not really testing all of the functionality of
Lazy. Some assertions were never executed

Remove SynchronizedLazy since it was unused

Implement Lazy.use() since this was only implemented for UnsafeLazy and
silently did nothing for LockingLazy

### CODE REVIEW

**Comment by big-guy** [comment]
@bot-gradle test this

File: subprojects/functional/src/main/java/org/gradle/internal/lazy/LockingLazy.java [hunk 1]
```diff
@@ -50,35 +53,19 @@ public LockingLazy(Supplier<T> supplier) {
 
     @Override
     public T get() {
-        return ensureValue();
-    }
-
-    private T ensureValue() {
```
Comment (big-guy) [inline subprojects/functional/src/main/java/org/gradle/internal/lazy/LockingLazy.java:56]: I actually tried to break this by running with a varying number of concurrent threads and couldn't get it to fail with Java 11 on macOS.  

It's possible this was good enough or only caused concurrency problems in specific combinations of OS/JVM/thread contention.

I trust that Guava's implementation is probably the best we could do, so I switched to that here.

**Review by DPUkyle — APPROVED** [review]
LGTM; glad that the tests couldn't reproduce any concurrency issues.

**Comment by DPUkyle** [comment]
@bot-gradle test and merge

**Comment by blindpirate** [comment]
@bot-gradle test and merge

**Pull request merged**